File: scripts/owned-process-emergency.mjs
import { readFile, readdir } from 'node:fs/promises'
import { setTimeout as nodeDelay } from 'node:timers/promises'

function parseStat(text) {
  const end = text.lastIndexOf(')')
  if (end < 0) throw new Error('malformed /proc stat')
  const fields = text.slice(end + 2).trim().split(/\s+/)
  if (fields.length < 20) throw new Error('malformed /proc stat fields')
  return {
    state: fields[0],
    ppid: Number(fields[1]),
    starttime: fields[19],
  }
}

export async function readProcessIdentity(pid) {
  try { return { pid, ...parseStat(await readFile(`/proc/${pid}/stat`, 'utf8')) } }
  catch (error) { if (error?.code === 'ENOENT') return null; throw error }
}

async function listProcessIds() {
  return (await readdir('/proc', { withFileTypes: true }))
    .filter(entry => entry.isDirectory() && /^\d+$/.test(entry.name))
    .map(entry => Number(entry.name))
}

function isStopped(identity) {
  return identity != null && /^[Tt]$/.test(identity.state)
}

async function defaultSnapshotDescendants(rootPid, dependencies) {
  const processes = []
  for (const pid of await dependencies.listProcessIds()) {
    const identity = await dependencies.readIdentity(pid)
    if (identity) processes.push({ pid, ...identity })
  }

  const descendants = []
  const seen = new Set([rootPid])
  const pending = [{ pid: rootPid, depth: 0 }]
  while (pending.length > 0) {
    const parent = pending.shift()
    for (const process of processes) {
      if (process.ppid !== parent.pid || seen.has(process.pid)) continue
      const descendant = { ...process, depth: parent.depth + 1 }
      descendants.push(descendant)
      seen.add(process.pid)
      pending.push(descendant)
    }
  }
  return descendants
}

async function waitUntil(check, timeoutMs, delay) {
  const deadline = Date.now() + timeoutMs
  while (true) {
    if (await check()) return true
    if (Date.now() >= deadline) return false
    await delay(Math.min(10, Math.max(1, deadline - Date.now())))
  }
}

async function requireMatching(identity, readIdentity) {
  const actual = await readIdentity(identity.pid)
  if (!actual || actual.starttime !== identity.starttime) {
    throw new Error(`owned process identity changed while containing PID ${identity.pid}`)
  }
  return actual
}

async function signalExact(identity, signal, dependencies) {
  await requireMatching(identity, dependencies.readIdentity)
  try { dependencies.signalProcess(identity.pid, signal) }
  catch (error) {
    if (error?.code === 'ESRCH') {
      throw new Error(`owned process identity disappeared before ${signal} for PID ${identity.pid}`, { cause: error })
    }
    throw error
  }
}

async function cleanupPartialFreeze(child, directIdentity, frozen, childExit, dependencies) {
  const errors = []
  for (const identity of [...frozen.values()].sort((left, right) => right.depth - left.depth)) {
    try {
      const actual = await dependencies.readIdentity(identity.pid)
      if (actual?.starttime === identity.starttime) dependencies.signalProcess(identity.pid, 'SIGKILL')
    } catch (error) { errors.push(error) }
  }
  try { child.kill('SIGKILL') } catch (error) { errors.push(error) }

  await Promise.race([childExit.catch(() => {}), dependencies.delay(dependencies.cleanupTimeoutMs)])

  for (const identity of frozen.values()) {
    try {
      const actual = await dependencies.readIdentity(identity.pid)
      if (actual?.starttime === identity.starttime && isStopped(actual)) dependencies.signalProcess(identity.pid, 'SIGCONT')
    } catch (error) { errors.push(error) }
  }
  if (directIdentity) {
    try {
      const actual = await dependencies.readIdentity(directIdentity.pid)
      if (actual?.starttime === directIdentity.starttime && isStopped(actual)) child.kill('SIGCONT')
    } catch (error) { errors.push(error) }
  }
  return errors
}

/**
 * Capability-scoped emergency shutdown for an unverified process-group anchor.
 * The owned direct ChildProcess is stopped first. Descendants are then frozen
 * by rechecked PID+starttime identities until a stopped-tree fixed point is
 * observed; only positive exact-identity SIGKILLs are used afterward.
 */
export async function freezeKillProcessTree(child, childExit, options = {}) {
  if (!child?.pid || !Number.isSafeInteger(child.pid) || child.pid <= 0) {
    throw new Error('owned direct child has no valid PID for emergency containment')
  }
  const dependencies = {
    listProcessIds,
    readIdentity: readProcessIdentity,
    signalProcess: process.kill.bind(process),
    delay: nodeDelay,
    maxFreezeIterations: 32,
    stopTimeoutMs: 500,
    exitTimeoutMs: 2_000,
    cleanupTimeoutMs: 250,
    ...options,
  }
  const frozen = new Map()
  let directIdentity

  try {
    if (!child.kill('SIGSTOP')) throw new Error('owned direct child could not be stopped')
    const direct = await dependencies.readIdentity(child.pid)
    if (!direct) throw new Error('owned direct child disappeared after SIGSTOP')
    directIdentity = { pid: child.pid, starttime: direct.starttime, depth: 0 }
    const directStopped = await waitUntil(async () => {
      const actual = await requireMatching(directIdentity, dependencies.readIdentity)
      return isStopped(actual)
    }, dependencies.stopTimeoutMs, dependencies.delay)
    if (!directStopped) throw new Error('owned direct child did not stop during emergency containment')

    let converged = false
    for (let iteration = 0; iteration < dependencies.maxFreezeIterations; iteration += 1) {
      const directActual = await requireMatching(directIdentity, dependencies.readIdentity)
      if (!isStopped(directActual)) throw new Error('owned direct child resumed during emergency containment')
      for (const identity of frozen.values()) {
        const actual = await requireMatching(identity, dependencies.readIdentity)
        if (!isStopped(actual)) throw new Error(`owned descendant PID ${identity.pid} resumed during emergency containment`)
      }

      const snapshot = await defaultSnapshotDescendants(child.pid, dependencies)
      const newlyDiscovered = snapshot
        .filter(identity => !frozen.has(identity.pid))
        .sort((left, right) => left.depth - right.depth)
      if (newlyDiscovered.length === 0) {
        converged = true
        break
      }

      for (const identity of newlyDiscovered) {
        await signalExact(identity, 'SIGSTOP', dependencies)
        frozen.set(identity.pid, identity)
        const stopped = await waitUntil(async () => {
          const actual = await requireMatching(identity, dependencies.readIdentity)
          return isStopped(actual)
        }, dependencies.stopTimeoutMs, dependencies.delay)
        if (!stopped) throw new Error(`owned descendant PID ${identity.pid} did not stop`)
      }
    }
    if (!converged) throw new Error(`owned process tree did not converge after ${dependencies.maxFreezeIterations} freeze iterations`)

    for (const identity of [...frozen.values()].sort((left, right) => right.depth - left.depth)) {
      await signalExact(identity, 'SIGKILL', dependencies)
    }
    const directBeforeKill = await requireMatching(directIdentity, dependencies.readIdentity)
    if (!isStopped(directBeforeKill)) throw new Error('owned direct child resumed before emergency kill')
    if (!child.kill('SIGKILL')) throw new Error('owned direct child could not be killed')

    const directExited = await Promise.race([
      childExit.then(() => true),
      dependencies.delay(dependencies.exitTimeoutMs).then(() => false),
    ])
    if (!directExited) throw new Error('owned direct child did not exit during emergency shutdown')

    const identities = [directIdentity, ...frozen.values()]
    const allGone = await waitUntil(async () => {
      for (const identity of identities) {
        const actual = await dependencies.readIdentity(identity.pid)
        if (actual?.starttime === identity.starttime) return false
        if (actual && actual.starttime !== identity.starttime) {
          throw new Error(`owned process PID ${identity.pid} was reused before shutdown verification`)
        }
      }
      return (await defaultSnapshotDescendants(child.pid, dependencies)).length === 0
    }, dependencies.exitTimeoutMs, dependencies.delay)
    if (!allGone) throw new Error('verified owned descendants remained after emergency shutdown')
  } catch (error) {
    const cleanupErrors = await cleanupPartialFreeze(child, directIdentity, frozen, childExit, dependencies)
    if (cleanupErrors.length > 0) {
      throw new AggregateError([error, ...cleanupErrors], `${error.message}; partial freeze cleanup also failed`)
    }
    throw error
  }
}

File: scripts/owned_process_emergency.py
#!/usr/bin/env python3
"""Fail-closed owned-tree cleanup using retained Linux pidfds."""

from __future__ import annotations

from dataclasses import dataclass
import json
import os
from pathlib import Path
import select
import signal
import sys
import time
from typing import Protocol

FRAME = 'SNL_OWNED_PROCESS_EMERGENCY_RESULT\t'


class EmergencyError(RuntimeError):
    pass


@dataclass(frozen=True)
class Identity:
    pid: int
    ppid: int
    starttime: str
    state: str


@dataclass
class Capability:
    pid: int
    ppid: int
    starttime: str
    depth: int
    fd: int


class Kernel(Protocol):
    def require_capabilities(self) -> None: ...
    def pidfd_open(self, pid: int) -> int: ...
    def read_identity(self, pid: int) -> Identity | None: ...
    def list_pids(self) -> list[int]: ...
    def send(self, fd: int, sig: int) -> None: ...
    def ready(self, fd: int, timeout: float) -> bool: ...
    def close(self, fd: int) -> None: ...
    def sleep(self, seconds: float) -> None: ...


class LinuxKernel:
    def require_capabilities(self) -> None:
        if not sys.platform.startswith('linux'):
            raise EmergencyError('pidfd emergency cleanup requires Linux')
        if not hasattr(os, 'pidfd_open') or not hasattr(signal, 'pidfd_send_signal'):
            raise EmergencyError('pidfd emergency cleanup capability is unavailable')

    def pidfd_open(self, pid: int) -> int:
        return os.pidfd_open(pid, 0)

    def read_identity(self, pid: int) -> Identity | None:
        try:
            text = Path(f'/proc/{pid}/stat').read_text()
        except OSError as error:
            if error.errno in (2, 3):
                return None
            raise
        end = text.rfind(')')
        if end < 0:
            raise EmergencyError(f'malformed /proc identity for PID {pid}')
        fields = text[end + 2:].split()
        if len(fields) < 20:
            raise EmergencyError(f'malformed /proc identity fields for PID {pid}')
        return Identity(pid=pid, state=fields[0], ppid=int(fields[1]), starttime=fields[19])

    def list_pids(self) -> list[int]:
        return [int(entry.name) for entry in Path('/proc').iterdir() if entry.name.isdigit()]

    def send(self, fd: int, sig: int) -> None:
        signal.pidfd_send_signal(fd, sig, None, 0)

    def ready(self, fd: int, timeout: float) -> bool:
        poller = select.poll()
        poller.register(fd, select.POLLIN)
        return bool(poller.poll(max(0, int(timeout * 1000))))

    def close(self, fd: int) -> None:
        os.close(fd)

    def sleep(self, seconds: float) -> None:
        time.sleep(seconds)


def _open_verified(pid: int, expected_starttime: str, kernel: Kernel) -> tuple[int, Identity]:
    try:
        fd = kernel.pidfd_open(pid)
    except ProcessLookupError as error:
        raise EmergencyError(f'owned process PID {pid} disappeared before pidfd acquisition') from error
    try:
        actual = kernel.read_identity(pid)
        if actual is None or actual.starttime != expected_starttime:
            raise EmergencyError(f'owned process identity mismatch for PID {pid}')
        return fd, actual
    except BaseException:
        kernel.close(fd)
        raise


def _snapshot(kernel: Kernel) -> dict[int, Identity]:
    result: dict[int, Identity] = {}
    for pid in kernel.list_pids():
        identity = kernel.read_identity(pid)
        if identity is not None:
            result[pid] = identity
    return result


def _descendant_depths(
    root: Capability,
    snapshot: dict[int, Identity],
    excluded_roots: set[int] | None = None,
) -> dict[int, int]:
    depths = {root.pid: 0}
    blocked = set(excluded_roots or ())
    changed = True
    while changed:
        changed = False
        for identity in snapshot.values():
            if identity.pid in depths or identity.pid in blocked:
                continue
            if identity.ppid in blocked:
                blocked.add(identity.pid)
                changed = True
                continue
            if identity.ppid not in depths:
                continue
            depths[identity.pid] = depths[identity.ppid] + 1
            changed = True
    return depths


def _wait_stopped(capability: Capability, timeout: float, kernel: Kernel) -> None:
    deadline = time.monotonic() + timeout
    while True:
        actual = kernel.read_identity(capability.pid)
        if actual is None or actual.starttime != capability.starttime:
            raise EmergencyError(f'owned process identity changed while freezing PID {capability.pid}')
        if actual.state in ('T', 't'):
            return
        if time.monotonic() >= deadline:
            raise EmergencyError(f'owned process PID {capability.pid} did not stop')
        kernel.sleep(0.005)


def _send_best_effort(capability: Capability, sig: int, kernel: Kernel) -> None:
    try:
        kernel.send(capability.fd, sig)
    except ProcessLookupError:
        pass


def _wait_all_ready(capabilities: list[Capability], timeout: float, kernel: Kernel) -> bool:
    deadline = time.monotonic() + timeout
    for capability in capabilities:
        if not kernel.ready(capability.fd, max(0.0, deadline - time.monotonic())):
            return False
    return True


def emergency_shutdown(
    root_pid: int,
    expected_starttime: str,
    expected_ppid: int | None,
    *,
    kernel: Kernel | None = None,
    max_freeze_iterations: int = 32,
    exit_timeout: float = 3.0,
    descendants_only: bool = False,
) -> dict[str, object]:
    """Freeze and kill an owned closure using only retained pidfds.

    ``descendants_only`` leaves the verified subreaper root alive and excludes
    this helper's branch, allowing the root to acknowledge completion over IPC.
    """
    kernel = kernel or LinuxKernel()
    kernel.require_capabilities()
    if root_pid <= 0 or not expected_starttime:
        raise EmergencyError('invalid owned root identity')

    capabilities: dict[int, Capability] = {}
    root: Capability | None = None
    excluded = {os.getpid()} if descendants_only and isinstance(kernel, LinuxKernel) else set()
    try:
        root_fd, root_identity = _open_verified(root_pid, expected_starttime, kernel)
        if expected_ppid is not None and root_identity.ppid != expected_ppid:
            kernel.close(root_fd)
            raise EmergencyError(f'owned root ancestry mismatch for PID {root_pid}')
        root = Capability(root_pid, root_identity.ppid, root_identity.starttime, 0, root_fd)
        capabilities[root_pid] = root
        if not descendants_only:
            kernel.send(root.fd, signal.SIGSTOP)
            _wait_stopped(root, 0.5, kernel)

        converged = False
        iteration = -1
        for iteration in range(max_freeze_iterations):
            snapshot = _snapshot(kernel)
            current_root = snapshot.get(root_pid)
            if current_root is None or current_root.starttime != root.starttime:
                raise EmergencyError(f'owned root identity changed while scanning PID {root_pid}')
            depths = _descendant_depths(root, snapshot, excluded)
            new_capabilities: list[Capability] = []
            for pid, depth in sorted(depths.items(), key=lambda item: item[1]):
                if pid == root_pid or pid in capabilities:
                    continue
                observed = snapshot[pid]
                parent = capabilities.get(observed.ppid)
                if parent is None:
                    continue
                parent_now = snapshot.get(parent.pid)
                if parent_now is None or parent_now.starttime != parent.starttime:
                    continue
                try:
                    fd, verified = _open_verified(pid, observed.starttime, kernel)
                except EmergencyError:
                    continue
                capability = Capability(pid, verified.ppid, verified.starttime, depth, fd)
                try:
                    if verified.state not in ('Z', 'X', 'x'):
                        kernel.send(fd, signal.SIGSTOP)
                        _wait_stopped(capability, 0.5, kernel)
                except BaseException:
                    kernel.close(fd)
                    raise
                capabilities[pid] = capability
                new_capabilities.append(capability)
            if not new_capabilities:
                converged = True
                break
            kernel.sleep(0.005)
        if not converged:
            raise EmergencyError(f'owned process tree did not converge after {max_freeze_iterations} freeze iterations')

        targets = [capability for pid, capability in capabilities.items() if not descendants_only or pid != root_pid]
        ordered = sorted(targets, key=lambda item: item.depth, reverse=True)
        for capability in ordered:
            _send_best_effort(capability, signal.SIGKILL, kernel)
        if not _wait_all_ready(ordered, exit_timeout, kernel):
            raise EmergencyError('retained owned process capabilities did not become ready after SIGKILL')
        return {'ok': True, 'retainedCapabilities': len(ordered), 'freezeIterations': iteration + 1}
    except BaseException:
        targets = [capability for pid, capability in capabilities.items() if not descendants_only or pid != root_pid]
        ordered = sorted(targets, key=lambda item: item.depth, reverse=True)
        for capability in ordered:
            _send_best_effort(capability, signal.SIGKILL, kernel)
        _wait_all_ready(ordered, min(exit_timeout, 0.5), kernel)
        raise
    finally:
        for capability in capabilities.values():
            try:
                kernel.close(capability.fd)
            except OSError:
                pass


def main(argv: list[str]) -> int:
    try:
        descendants_only = False
        if argv[-1:] == ['--descendants-only']:
            descendants_only = True
            argv = argv[:-1]
        if len(argv) not in (3, 4):
            raise EmergencyError('usage: owned_process_emergency.py PID STARTTIME [EXPECTED_PPID] [--descendants-only]')
        result = emergency_shutdown(
            int(argv[1]), argv[2], int(argv[3]) if len(argv) == 4 else None,
            descendants_only=descendants_only,
        )
        print(FRAME + json.dumps(result, separators=(',', ':')), flush=True)
        return 0
    except BaseException as error:
        result = {'ok': False, 'cleanupIncomplete': True, 'message': str(error)}
        print(FRAME + json.dumps(result, separators=(',', ':')), flush=True)
        return 1


if __name__ == '__main__':
    raise SystemExit(main(sys.argv))
